java/src/main/native/src/jni_writer_data_sink.hpp

```diff
@@ -0,0 +1,177 @@
+    // is likely being triggered by the C++ runtime during shutdown. In that case the JVM may
+    // already be destroyed and this thread should not try to attach to get an environment.
+    JNIEnv *env = nullptr;
+    if (jvm->GetEnv(reinterpret_cast<void **>(&env), cudf::jni::MINIMUM_JNI_VERSION) == JNI_OK) {
+      env->DeleteGlobalRef(callback);
+      if (current_buffer != nullptr) {
+        env->DeleteGlobalRef(current_buffer);
+      }
+    }
+    callback = nullptr;
+    current_buffer = nullptr;
+  }
+
+  void host_write(void const *data, size_t size) override {
+    JNIEnv *env = cudf::jni::get_jni_env(jvm);
+    long left_to_copy = static_cast<long>(size);
+    const char *copy_from = static_cast<const char *>(data);
+    while (left_to_copy > 0) {
+      long buffer_amount_available = current_buffer_len - current_buffer_written;
+      if (buffer_amount_available <= 0) {
+        // should never be < 0, but just to be safe
+        rotate_buffer(env);
+        buffer_amount_available = current_buffer_len - current_buffer_written;
+      }
+      long amount_to_copy =
+          left_to_copy < buffer_amount_available ? left_to_copy : buffer_amount_available;
+      char *copy_to = current_buffer_data + current_buffer_written;
+
+      std::memcpy(copy_to, copy_from, amount_to_copy);
+      copy_from = copy_from + amount_to_copy;
+      current_buffer_written += amount_to_copy;
+      total_written += amount_to_copy;
+      left_to_copy -= amount_to_copy;
+    }
+  }
+
+  bool supports_device_write() const override { return true; }
+
+  void device_write(void const *gpu_data, size_t size, rmm::cuda_stream_view stream) override {
+    JNIEnv *env = cudf::jni::get_jni_env(jvm);
+    long left_to_copy = static_cast<long>(size);
+    const char *copy_from = static_cast<const char *>(gpu_data);
+    while (left_to_copy > 0) {
+      long buffer_amount_available = current_buffer_len - current_buffer_written;
+      if (buffer_amount_available <= 0) {
+        // should never be < 0, but just to be safe
+        stream.synchronize();
+        rotate_buffer(env);
+        buffer_amount_available = current_buffer_len - current_buffer_written;
+      }
+      long amount_to_copy =
+          left_to_copy < buffer_amount_available ? left_to_copy : buffer_amount_available;
+      char *copy_to = current_buffer_data + current_buffer_written;
+
+      CUDF_CUDA_TRY(cudaMemcpyAsync(copy_to, copy_from, amount_to_copy, cudaMemcpyDeviceToHost,
+                                    stream.value()));
+
+      copy_from = copy_from + amount_to_copy;
+      current_buffer_written += amount_to_copy;
+      total_written += amount_to_copy;
+      left_to_copy -= amount_to_copy;
+    }
+    stream.synchronize();
+  }
+
+  std::future<void> device_write_async(void const *gpu_data, size_t size,
+                                       rmm::cuda_stream_view stream) override {
+    // Call the sync version until figuring out how to write asynchronously.
```

Comment:
I would have expected that the sync version called the async version and then synchronized. Is there anything else needed here to make it async outside of calling `cudaMemcpyAsync`? If the call to write takes a long time to loop through the buffers, I could see spawning a thread to do the work. In that case I would think a common function that could be called and then sync'd or passed to the `std::async` function.

Reply:
> Is there anything else needed here to make it async outside of calling cudaMemcpyAsync?

Based on the signature, I *think* the `async`ness is not cudaMemcpy async-ness, but `std::thread` async-ness.
It's probably valid for the `async` function to call the sync version via `std::async`. I'll add this to the follow-up issue.

Reply:
#12456.

Reply:
I was under the impression that this was new code. I'm perfectly fine with filing an issue even if the issue number is suspect.